applications/FluidDynamicsApplication/custom_conditions/adjoint_monolithic_wall_condition.cpp

```diff
@@ -0,0 +1,779 @@
+    }
+}
+
+template<unsigned int TDim, unsigned int TNumNodes>
+void AdjointMonolithicWallCondition<TDim, TNumNodes>::GetFirstDerivativesVector(
+    Vector& Values,
+    int Step) const
+{
+    if (Values.size() != TFluidLocalSize) {
+        Values.resize(TFluidLocalSize, false);
+    }
+
+    Values.clear();
+}
+
+template<>
+void AdjointMonolithicWallCondition<2, 2>::GetSecondDerivativesVector(
+    Vector& Values,
+    int Step) const
+{
+    if (Values.size() != 6) {
+        Values.resize(6, false);
+    }
+
+    IndexType local_index = 0;
+    for (IndexType i = 0; i < 2; ++i) {
+        const auto& r_node = this->GetGeometry()[i];
+        Values[local_index++] = r_node.FastGetSolutionStepValue(ADJOINT_FLUID_VECTOR_3_X);
+        Values[local_index++] = r_node.FastGetSolutionStepValue(ADJOINT_FLUID_VECTOR_3_Y);
+        Values[local_index++] = 0.0;
+    }
+}
+
+template<>
+void AdjointMonolithicWallCondition<3, 3>::GetSecondDerivativesVector(
+    Vector& Values,
+    int Step) const
+{
+    if (Values.size() != 12) {
+        Values.resize(12, false);
+    }
+
+    IndexType local_index = 0;
+    for (IndexType i = 0; i < 3; ++i) {
+        const auto& r_node = this->GetGeometry()[i];
+        Values[local_index++] = r_node.FastGetSolutionStepValue(ADJOINT_FLUID_VECTOR_3_X);
+        Values[local_index++] = r_node.FastGetSolutionStepValue(ADJOINT_FLUID_VECTOR_3_Y);
+        Values[local_index++] = r_node.FastGetSolutionStepValue(ADJOINT_FLUID_VECTOR_3_Z);
+        Values[local_index++] = 0.0;
+    }
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+int AdjointMonolithicWallCondition<TDim, TNumNodes>::Check(
+        const ProcessInfo& rCurrentProcessInfo) const
+{
+    return 0;
+}
+
+template<>
+double AdjointMonolithicWallCondition<2, 2>::CalculateDetJ(
+    const double Area) const
+{
+    return Area * 0.5;
+}
+
+template<>
+double AdjointMonolithicWallCondition<3, 3>::CalculateDetJ(
+    const double Area) const
+{
+    return Area * 2.0;
+}
+
+template<>
+void AdjointMonolithicWallCondition<2, 2>::CalculateDetJShapeDerivatives(
+    BoundedVector<double, TCoordsLocalSize>& rDetJDerivatives,
+    const BoundedVector<double, TCoordsLocalSize>& rAreaDerivatives) const
+{
+    noalias(rDetJDerivatives) = rAreaDerivatives * 0.5;
+}
+
+template<>
+void AdjointMonolithicWallCondition<3, 3>::CalculateDetJShapeDerivatives(
+    BoundedVector<double, TCoordsLocalSize>& rDetJDerivatives,
+    const BoundedVector<double, TCoordsLocalSize>& rAreaDerivatives) const
+{
+    noalias(rDetJDerivatives) = rAreaDerivatives * 2.0;
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+void AdjointMonolithicWallCondition<TDim, TNumNodes>::CalculateData(
+    double& rArea,
+    double& rDetJ,
+    array_1d<double, TDim>& rUnitNormal) const
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(!this->Has(NORMAL))
+        << "NORMAL is not defined for " << this->Info() << ".\n";
+
+    const auto& aux_normal = this->GetValue(NORMAL);
```

Comment:
Even though this should be the proper usage of the normal, traditionally we've stored in the historical database. Just mentioning because this could lead to hard to track errors.

Reply:
this is the condition `NORMAL` which is the actual normal which we calculate. That's why I'm using it to avoid problems at the corners :)

I understand your point. So I kept usage of nodal historical NORMAL only for the `SLIP` matrix and vector rotation (which is unavoidable to get rid of). Everything else is based on condition NORMAL.